### PR TITLE
Add support for the MEME psp format

### PR DESCRIPTION
**Pull request by gregvonkuster**
Ping @bgruening - this is in preparation for the tool you requested - http://meme-suite.org/doc/psp-gen.html

### CODE REVIEW

File: config/datatypes_conf.xml.sample [hunk 1]
```diff
@@ -269,6 +269,7 @@
     <datatype extension="obo" type="galaxy.datatypes.text:Obo" mimetype="text/html" display_in_upload="true" />
     <datatype extension="owl" type="galaxy.datatypes.xml:Owl" mimetype="text/html" display_in_upload="true" />
     <datatype extension="png" type="galaxy.datatypes.images:Png" mimetype="image/png"/>
+    <datatype extension="psp" type="galaxy.datatypes.sequence:Psp" display_in_upload="true" description="The MEME Position Specific Priors (PSP) format includes the name of the sequence for which a prior distribution corresponds." description_url="http://meme-suite.org/doc/psp-format.html">
```
Comment (jmchilton) [inline config/datatypes_conf.xml.sample]: Need to close the datatype tag here.

File: lib/galaxy/datatypes/sequence.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import re
 import string
 import sys
+from __builtin__ import False
```
Comment (bgruening) [inline lib/galaxy/datatypes/sequence.py]: Why is this needed?

File: lib/galaxy/datatypes/sequence.py [hunk 3]
```diff
@@ -1208,3 +1210,65 @@ def sniff(self, filename):
             pass
 
         return False
+
+
+class Psp(Sequence):
+    """Class representing MEME Position Specific Priors"""
+    file_ext = "psp"
```
Comment (bgruening) [inline lib/galaxy/datatypes/sequence.py]: Should we call it meme_psp and the class MemePsp?

File: lib/galaxy/datatypes/sequence.py [hunk 4]
```diff
@@ -1208,3 +1210,65 @@ def sniff(self, filename):
...
+        """
+        def floats_verified(l):
+            for item in l.split():
+                try:
+                    float(item)
+                except:
+                    return False
+            return True
+        try:
+            fh = open(filename)
```
Comment (bgruening) [inline lib/galaxy/datatypes/sequence.py]: can we use `with open(filename) as fh:`

File: lib/galaxy/datatypes/sequence.py [hunk 5]
```diff
@@ -1208,3 +1210,65 @@ def sniff(self, filename):
...
+                if not line:
+                    # EOF.
+                    break
+                line = line.strip()
+                if line:
+                    if line.startswith( '>' ):
+                        # The line must not be blank, nor start with '>'
+                        line = fh.readline().strip()
+                        if line == '' or line.startswith( '>' ):
+                            break
```
Comment (bgruening) [inline lib/galaxy/datatypes/sequence.py]: return False?

File: lib/galaxy/datatypes/sequence.py [hunk 6]
```diff
@@ -1208,3 +1210,65 @@ def sniff(self, filename):
...
+                        # All items within the line must be floats.
+                        if not floats_verified(line):
+                            break
+                        # If there is a second line within the ID section,
+                        # all items within the line must be floats.
+                        line = fh.readline().strip()
+                        if line:
+                            if not floats_verified(line):
+                                break
```
Comment (bgruening) [inline lib/galaxy/datatypes/sequence.py]: return False

File: lib/galaxy/datatypes/sequence.py [hunk 7]
```diff
@@ -1208,3 +1210,65 @@ def sniff(self, filename):
...
+        def floats_verified(l):
+            for item in l.split():
+                try:
+                    float(item)
+                except:
+                    return False
+            return True
+        try:
+            fh = open(filename)
+            while True:
```
Comment (bgruening) [inline lib/galaxy/datatypes/sequence.py]: We should not iterate through the entire file, maybe the top 100 lines?

**Review by bgruening — COMMENTED** [review]
Thanks @gregvonkuster!

File: lib/galaxy/datatypes/sequence.py (hunk 2 above)
Comment (gregvonkuster) [reply lib/galaxy/datatypes/sequence.py]: Not sure - I didn't add it.  I just changed the import order so that flake8 would be happy.

Comment (nsoranzo) [reply lib/galaxy/datatypes/sequence.py]: But that import doesn't appear anywhere else.

File: config/datatypes_conf.xml.sample [hunk 8]
```diff
@@ -691,6 +692,7 @@
     <sniffer type="galaxy.datatypes.triples:Jsonld"/>
     <sniffer type="galaxy.datatypes.sequence:Maf"/>
     <sniffer type="galaxy.datatypes.sequence:Lav"/>
+    <sniffer type="galaxy.datatypes.sequence:Psp"/>
```
Comment (martenson) [inline config/datatypes_conf.xml.sample]: should this read `MemePsp` too?

**Review by bgruening — APPROVED** [review]
Thanks Greg!

**Comment by dannon** [comment]
Failing tests are unrelated to this.

**Comment by dannon** [comment]
Just realized the merge here will break our *very* new  (overnight) lint guidelines, will open up a subsequent PR to fix the styling.

File: lib/galaxy/datatypes/sequence.py [hunk 9]
```diff
@@ -1208,3 +1209,67 @@ def sniff(self, filename):
...
+            for item in l.split():
+                try:
+                    float(item)
+                except:
+                    return False
+            return True
+        try:
+            num_lines = 0
+            with open(filename) as fh:
+                line = fh.readline()
```
Comment (jmchilton) [inline lib/galaxy/datatypes/sequence.py:1245]: Please limit these readlines in future - this will use an unbounded amount of memory on certain inputs.
